sbc_compassion/models/correspondence_s2b_generator.py

```diff
@@ -189,42 +189,54 @@ def generate_letters_job(self, utms=None):
         :return: True
         """
-        if utms is None:
-            utms = dict()
-        utms = defaultdict(lambda: None, utms)
-        letters = self.env["correspondence"]
-        for sponsorship in self.sponsorship_ids:
-            text = self._get_text(sponsorship)
-            vals = {
-                "sponsorship_id": sponsorship.id,
-                "store_letter_image": False,
-                "template_id": self.template_id.id,
-                "direction": "Supporter To Beneficiary",
-                "source": self.source,
-                "original_language_id": self.language_id.id,
-                "original_text": text,
-                "campaign_id": utms["campaign"],
-                "medium_id": utms["medium"],
-                "source_id": utms["source"],
-            }
-            if self.image_ids:
-                vals["original_attachment_ids"] = [
-                    (
-                        0,
-                        0,
-                        {
-                            "datas": atchmt.datas,
-                            "name": atchmt.name,
-                            "res_model": letters._name,
-                        },
-                    )
-                    for atchmt in self.image_ids
-                ]
-            letters += letters.create(vals)
+        try:
+            if utms is None:
+                utms = dict()
+            utms = defaultdict(lambda: None, utms)
+            letters = self.env["correspondence"]
+            for sponsorship in self.sponsorship_ids:
+                text = self._get_text(sponsorship)
+                vals = {
+                    "sponsorship_id": sponsorship.id,
+                    "store_letter_image": False,
+                    "template_id": self.template_id.id,
+                    "direction": "Supporter To Beneficiary",
+                    "source": self.source,
+                    "original_language_id": self.language_id.id,
+                    "original_text": text,
+                    "campaign_id": utms["campaign"],
+                    "medium_id": utms["medium"],
+                    "source_id": utms["source"],
+                }
+                if self.image_ids:
+                    vals["original_attachment_ids"] = [
+                        (
+                            0,
+                            0,
+                            {
+                                "datas": atchmt.datas,
+                                "name": atchmt.name,
+                                "res_model": letters._name,
+                            },
+                        )
+                        for atchmt in self.image_ids
+                    ]
+                letters += letters.create(vals)
+
+            self.letter_ids = letters
+
+            # If the operation succeeds, notify the user
+            message = "Letters have been successfully generated."
+            self.env.user.notify_success(message=message)
+
```

Comment:
The `write` at line 237 should be moved at the end of the try block, otherwise we will write the state of the generation to done even when an error occured.

Reply:
The `write` at line 237 has been moved at the end of the try block to prevent the write state of the generation to done when an error occurs.